designs/2018-processors-improvements/README.md

```diff
@@ -0,0 +1,284 @@
+
+1. Currently, application of processors happens inside of the `Linter#verify()` and `Linter#verifyAndFix()` methods. In order to implement this proposal, application of processors must happen outside of `Linter` because configuration calculation happens outside of `Linter`.
+2. The application of processors must move into `CLIEngine#executeOnText()` and `CLIEngine#executeOnFiles()` to work alongside with the current configuration calculation. That also means `CLIEngine` becomes responsible for recombining and postprocessing results from multiple `Linter#verify()` calls.
+3. A new option called `disableFixes` will be added to the third argument of `Linter#verify()`. When `disableFixes` is set to `true`, `Linter#verify()` will not produce fix information for autofixable rules. This is necessary to turn off autofixing in case one of the processors does not support autofixing.
+
+A high-level overview of these changes is:
+
+```js
+// inside of cli-engine.js
+function verify(filename, code) {
+
+    //....(resolve config)....
+
+    return postprocess(
+        preprocess(code, filename).map(item => {
+
+            // avoid infinite loop
+            if ((typeof item === "string") ||
+                (path.extname(filename) === path.extname(item.filename))
+            ) {
+                return linter.verify(code, config, {
+                    ...options,
+                    filePath: filename,
+                    disableFixes: !supportsAutofix
+                })
+            }
+
+            // Recursive call
+            return verify(item.filename, item.code)
+
+        }),
+        filename,
+    )
+}
+```
+
+With these changes, the existing `Linter` options of `preprocess` and `postprocess` will remain in place so the browser API will not break. The ESLint CLI, however, will not use these options. (See Backwards Compatibility Analysis for more information.)
+
+## Documentation
+
+The following areas of the website will need to be updated:
+
+* [Processors in Plugins](https://eslint.org/docs/developer-guide/working-with-plugins#processors-in-plugins)
+* [Configuring ESLint](https://eslint.org/docs/user-guide/configuring#configuring-eslint)
+
+Additionally, a blog post announcement should be made both to publicize the change and to encourage plugin owners to update their processors to be named instead of extension-named.
+
+## Drawbacks
+
+There are some drawbacks to this proposal, namely:
+
+1. Existing plugins would have to be updated so they can be used in this way.
+1. Existing configurations would have to be updated to make use of this design.
+
+Both of these drawbacks create a barrier to adoption and require changes by the end-user in order to take advantage of this design.
+
+## Backwards Compatibility Analysis
+
+This proposal is 100% backwards compatible until we remove the old way of defining processors. Both named and extension-based processors can be defined in the same plugin, such as:
```

Comment:
I've spent a while trying to figure this out, so please correct me if I've missed something. I think that this will require a major version bump when plugins start to support this API.

Currently, as well as after this change in the case of extension-named processors, code blocks inherit their parent file's filename. For example, `project/README.md` would contain blocks named `project/README.md`. In the Markdown plugin's readme right now, we [recommend configuring rules etc. using `*.md` in `overrides`](https://github.com/eslint/eslint-plugin-markdown#project-or-directory-wide-overrides-for-code-snippets).

When a processor is upgraded to support the named processor API, those code blocks would become `project/README.md/0.js`, and any rules configured for `*.md` files would no longer be applied to the code blocks inside of them.

Individually, the `eslint` adding support for this API won't break the existing (now called extension-named) processors, and `eslint-plugin-markdown` can start exporting both extension-named and named processors without breaking versions of ESLint that don't support the new API. But taken together, a version of ESLint that supports extension-named processors and an extension-named processor could require me to modify my configuration if I'm specifically targeting code blocks returned by a processor.

Am I lost in the trees here?

Reply:
That is true, but only if the processor is returning filenames with the code blocks. If the processor is still returning strings, then the code blocks get the same name as the parent file. And when ESLint is given string code blocks, it would stop matching against the code block filenames because that would be an infinite loop.

So, it would be a breaking change for the *plugin* if it changes from returning strings to returning filenames, but not a breaking change for ESLint.

Reply:
That makes sense. I’m happy to use the Markdown plugin to test the recommended upgrade path for processors once this is available in ESLint.